google_workspace_add_on/src/delta-sharing-client.js

```diff
@@ -0,0 +1,95 @@
+export class DeltaSharingClient {
+    constructor(profile) {
+        this.client = new DeltaSharingRestClient(profile);
+    }
+
+    listShares() {
+        let pageToken = null;
+        let shares = [];
+        do {
+            let resp = this.client.listShares(pageToken);
+            if ('items' in resp) {
+                shares.push(...resp.items);
+            }
+            pageToken = 'pageToken' in resp ? resp.pageToken : null;
+        } while (pageToken && pageToken != '');
+        return shares;
+    }
+
+    listAllTables(share) {
+        let pageToken = null;
+        let tables = [];
+        do {
+            let resp = this.client.listAllTables(share, pageToken);
+            if ('items' in resp) {
+                tables.push(...resp.items);
+            }
+            pageToken = 'pageToken' in resp ? resp.pageToken : null;
+        } while (pageToken && pageToken != '');
+        return tables;
+    }
+
+    queryTable(tableItem, limitHint) {
+        return this.client.queryTable(tableItem.share, tableItem.schema, tableItem.name, limitHint);
+    }
+}
+
+class DeltaSharingRestClient {
+    constructor(profile) {
+        this.profile = profile;
+    }
+
+    callEndpoint(url, method, payload) {
+        let headers = {
+            'Authorization': `Bearer ${this.profile.bearerToken}`
+        };
+        // Query table requires this header as well.
+        if (method == 'post') {
+            headers['Content-Type'] = 'application/json; charset=utf-8';
+        }
+        let payloadStr = payload ? JSON.stringify(payload) : null;
+        let params = {
+            method,
+            payload: payloadStr,
+            headers
+        };
+        return UrlFetchApp.fetch(
```

Comment:
Looks like `UrlFetchApp` is synchronized?

Reply:
Yes it is. Everything is synchronized in Google Apps Script. Asynchronicity is a lie!